library/bn_mul.h

```diff
@@ -666,55 +675,63 @@
  * x !=0, which we can detect using __OPTIMIZE__ (which is also defined by
  * clang and armcc5 under the same conditions).
  *
- * So, only use the optimized assembly below for optimized build, which avoids
- * the build error and is pretty reasonable anyway.
+ * If gcc needs to use r7, we use r1 as a scratch register and have a few extra
+ * instructions to preserve/restore it; otherwise, we can use r7 and avoid
+ * the preserve/restore overhead.
  */
-#if defined(__GNUC__) && !defined(__OPTIMIZE__)
-#define MULADDC_CANNOT_USE_R7
-#endif
-
-#if defined(__arm__) && !defined(MULADDC_CANNOT_USE_R7)
-
-#if defined(__thumb__) && !defined(__thumb2__)
+#define MULADDC_SCRATCH         "RS .req r1         \n\t"
+#define MULADDC_PRESERVE_R1     "mov    r10, r1     \n\t"
+#define MULADDC_RESTORE_R1      "mov    r1, r10     \n\t"
```

Comment:
Aren't these `MULADDC_PRESERVE_SCRATCH` and `MULADDC_RESTORE_SCRATCH`? (Possibly with "if necessary", or "opt[ional]")

Reply:
No, it's (maybe) preserving/restoring R1 so that it can be used as scratch.

Reply:
Yes, as they are implemented, but if a modification was to use `r0` as scratch instead, it would be `r0` that needed to be saved and restored, and as it stands this would also require the macro names to be changed, which seems to defeat the point of the abstraction